src/mappings/exchange.ts

```diff
@@ -58,873 +72,965 @@ export function handleTokenPurchase(event: TokenPurchase): void {
+      exchangeDayData.exchangeAddress = event.address
+      exchangeDayData.ethBalance = BigDecimal.fromString('0')
+      exchangeDayData.tokenBalance = BigDecimal.fromString('0')
+      exchangeDayData.marginalEthRate = BigDecimal.fromString('0')
+      exchangeDayData.ethVolume = BigDecimal.fromString('0')
+      exchangeDayData.ROI = BigDecimal.fromString('0')
+      exchangeDayData.totalEvents = BigInt.fromI32(0)
+      exchangeDayData.tokenPriceUSD = BigDecimal.fromString('0')
+    }
+    exchangeDayData.ethBalance = exchange.ethBalance
+    exchangeDayData.tokenBalance = exchange.tokenBalance
+    if (equalToZero(exchange.ethBalance)) {
+      // do nothing
+    } else {
+      exchangeDayData.marginalEthRate = exchange.tokenBalance.div(exchange.ethBalance).truncate(8)
+    }
+    exchangeDayData.ethVolume = exchangeDayData.ethVolume.plus(ethAmount)
+    exchangeDayData.ROI = exchange.ROI
+    exchangeDayData.tokenPriceUSD = exchange.priceUSD
+    exchangeDayData.totalEvents = exchangeDayData.totalEvents.plus(BigInt.fromI32(1))
+    exchangeDayData.save()
   }
-  exchangeDayData.ethVolume = exchangeDayData.ethVolume.plus(ethAmount)
-  exchangeDayData.ROI = exchange.ROI
-  exchangeDayData.tokenPriceUSD = exchange.priceUSD
-  exchangeDayData.totalEvents = exchangeDayData.totalEvents.plus(BigInt.fromI32(1))
-  exchangeDayData.save()
 }
 
 export function handleEthPurchase(event: EthPurchase): void {
   /****** Update Exchange ******/
   const exchangeID = event.address.toHex()
   const exchange = Exchange.load(exchangeID)
-  const ethAmount = event.params.eth_bought.toBigDecimal().div(exponentToBigDecimal(18))
-  let tokenAmount: BigDecimal
-  if (exchange.tokenDecimals == null) {
-    tokenAmount = event.params.tokens_sold.toBigDecimal()
-  } else {
-    tokenAmount = event.params.tokens_sold.toBigDecimal().div(exponentToBigDecimal(exchange.tokenDecimals))
-  }
-  exchange.ethBalance = exchange.ethBalance.minus(ethAmount)
-  exchange.tokenBalance = exchange.tokenBalance.plus(tokenAmount)
-  exchange.sellTokenCount = exchange.sellTokenCount.plus(BigInt.fromI32(1))
-  exchange.lastPrice = exchange.price
-
-  // Here we must handle div by zero, because someone could have bought all the eth or all the tokens
-  if (exchange.ethBalance.equals(BigDecimal.fromString('0'))) {
-    exchange.price = BigDecimal.fromString('0')
-  } else {
-    exchange.price = exchange.tokenBalance.div(exchange.ethBalance).truncate(18)
-    exchange.combinedBalanceInEth = exchange.ethBalance.plus(exchange.tokenBalance.div(exchange.price)).truncate(18)
-  }
+  if (exchange !== null) {
+    const ethAmount = event.params.eth_bought.toBigDecimal().div(exponentToBigDecimal(18))
+    let tokenAmount: BigDecimal
+    if (exchange.tokenDecimals == null) {
+      tokenAmount = event.params.tokens_sold.toBigDecimal()
+    } else {
+      tokenAmount = event.params.tokens_sold.toBigDecimal().div(exponentToBigDecimal(exchange.tokenDecimals))
```

Comment:
Do the same as explained in the above comment, expect also pass exchange.TokenDecimals as a parameter:
```
function convertTokenToDecimal (tokenAmount: BigInt, exchangeDecimals: i32): BigDecimal {
    tokenAmount.toBigDecimal().div(exponentToBigDecimal(exchangeDecimals))
}
```